core/go/internal/statemgr/domain_context.go

```diff
@@ -249,6 +258,26 @@ func (dc *domainContext) mergeInMemoryMatches(schema components.Schema, states [
 
 }
 
+func (dc *domainContext) GetStatesByID(dbTX *gorm.DB, schemaID tktypes.Bytes32, ids []string) (components.Schema, []*pldapi.State, error) {
+	idsAny := make([]any, len(ids))
+	for i, id := range ids {
+		idsAny[i] = id
+	}
+	query := query.NewQueryBuilder().In(".id", idsAny).Sort(".created").Query()
+	schema, matches, err := dc.ss.findStates(dc, dbTX, dc.domainName, &dc.contractAddress, schemaID, query, pldapi.StateStatusAll)
+	if err == nil {
+		var memMatches []*components.StateWithLabels
+		memMatches, err = dc.mergeUnFlushed(schema, matches, query, false)
```

Comment:
It looks like this function still filters out spent states - so there's a small chance we'll miss states that were created _and_ spent in memory before being flushed to the database.

Reply:
Good spot - I've updated so it does not exclude those locked-for-spend states